src/bson.rs

```diff
@@ -304,59 +304,108 @@ impl From<Value> for Bson {
         match a {
             Value::Number(x) => x
                 .as_i64()
-                .map(Bson::from)
+                .map(|i| {
+                    if i >= i32::MIN as i64 && i <= i32::MAX as i64 {
+                        Bson::I32(i as i32)
+                    } else {
+                        Bson::I64(i)
+                    }
+                })
                 .or_else(|| x.as_u64().map(Bson::from))
                 .or_else(|| x.as_f64().map(Bson::from))
                 .unwrap_or_else(|| panic!("Invalid number value: {}", x)),
             Value::String(x) => x.into(),
             Value::Bool(x) => x.into(),
             Value::Array(x) => Bson::Array(x.into_iter().map(Bson::from).collect()),
-            Value::Object(x) => {
-                Bson::from_extended_document(x.into_iter().map(|(k, v)| (k, v.into())).collect())
-            }
+            Value::Object(x) => Bson::from_extended_document(
+                x.into_iter().map(|(k, v)| (k, Bson::from(v))).collect(),
+            ),
             Value::Null => Bson::Null,
         }
     }
 }
 
 impl From<Bson> for Value {
     fn from(bson: Bson) -> Self {
-        match bson {
+        bson.into_relaxed_extjson()
+    }
+}
+
+impl Bson {
+    pub fn into_relaxed_extjson(self) -> Value {
+        match self {
+            Bson::FloatingPoint(v) if v.is_nan() => {
+                let s = if v.is_sign_negative() { "-NaN" } else { "NaN" };
```

Comment:
I figured it was more straightforward to manually implement the stringification of +/-Nan and Infinity rather than trying to rely on the default formatting that Rust gives